consensus/service/config/src/signer_set.rs

```diff
@@ -0,0 +1,351 @@
+    }
+}
+
+// The two types of signers we support.
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(tag = "type", content = "identity")]
+enum Signer {
+    Single(String),
+    Multi(SignerSet),
+}
+
+/// Wrapper around `mc_crypto_multisig::SignerSet` that allows us specifying
+/// singers via a string name
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+struct SignerSet {
+    threshold: u32,
+    signers: Vec<Signer>,
+}
+
+impl SignerSet {
+    /// Given a map of signer names to public keys, return a
+    /// `mc_crypto_multisig::SignerSet` that represents the same signing
+    /// structure that is described by this `SignerSet`.
+    pub fn resolve(
+        &self,
+        signer_identities: &HashMap<String, PemEd25519Public>,
+    ) -> Result<mc_crypto_multisig::SignerSet<Ed25519Public>, String> {
+        if self.threshold == 0 || self.threshold as usize > self.signers.len() {
+            return Err(format!(
+                "SignerSet contains invalid threshold of {}/{}",
+                self.threshold,
+                self.signers.len()
+            ));
+        }
+
+        let mut individual_signers = Vec::new();
+        let mut multi_signers = Vec::new();
+
+        for signer in self.signers.iter() {
+            match signer {
+                Signer::Single(identity) => {
+                    let signer = signer_identities
+                        .get(identity)
+                        .ok_or_else(|| format!("Unknown identity: {}", identity))?;
+                    individual_signers.push(signer.0);
+                }
+
+                Signer::Multi(signer_set) => {
+                    multi_signers.push(signer_set.resolve(signer_identities)?);
+                }
+            };
+        }
+
+        let signer_set =
+            mc_crypto_multisig::SignerSet::new(individual_signers, multi_signers, self.threshold);
+        if !signer_set.is_valid() {
+            return Err("SignerSet is invalid".into());
+        }
+        Ok(signer_set)
+    }
+}
+
+/// Signer set configuration, structured in a way that makes it a bit more sane
+/// for humans to review and edit.
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct SignerSetConfig {
+    /// A map of human readable name to a PEM-encoded public key, representing a
+    /// single signing identity.
+    #[serde_as(as = "HashMap<DisplayFromStr, DisplayFromStr>")]
+    signer_identities: HashMap<String, PemEd25519Public>,
+
+    /// Signer set configuration.
+    signer_set: SignerSet,
+}
+
+impl TryFrom<&SignerSetConfig> for mc_crypto_multisig::SignerSet<Ed25519Public> {
+    type Error = String;
+
+    fn try_from(signer_set_config: &SignerSetConfig) -> Result<Self, Self::Error> {
+        signer_set_config
+            .signer_set
+            .resolve(&signer_set_config.signer_identities)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_happy_flow() {
```

Comment:
Thank you! Agreed it would've been nice to have less repetition there but I favor making it easier to understand and provide a reference file over coming up with some tricks to shorten it.